Initial commit: Pyccel Cuda

bauom opened the pull request.

bauom committed:
added kernel class and cudaext.py

bauom committed:
added kernel class and cudaext.py

bauom committed:
remove prints

bauom committed:
added printers for arraycreation and thread indexing && added memory_location attr to variable

bauom committed:
fixed a cwarapper problem for cuda

bauom committed:
fixed cudadevicesynchronize call

bauom committed:
activate github action test

bauom committed:
remove unnecessary lines

bauom committed:
Merge branch 'master' of https://github.com/pyccel/pyccel into cuda_00

bauom committed:
Revert "remove unnecessary lines"

This reverts commit 8290b6026fa049df95ae12cc9ee4de01e5e6e239.

bauom committed:
fixed conflicts after merge with master

bauom committed:
fixed cudanameclashchecker

bauom committed:
fixed cudanameclasher typo

bauom committed:
Cuda tests

bauom committed:
added cuda_install action

bauom committed:
using local installer

bauom committed:
use docker container

bauom committed:
use docker container

bauom committed:
use docker container devel

bauom committed:
use docker container devel

bauom committed:
removed sudo from linux_install

bauom committed:
added y option to apt-get install cmd

bauom committed:
tzdata waiting for input

bauom committed:
added TIMEZONE

bauom committed:
installing python3 in container

bauom committed:
python-is-python3

bauom committed:
python-is-python3

bauom committed:
python-is-python3

bauom committed:
python-is-python3

bauom committed:
python-is-python3

bauom committed:
Merge branch 'master' of https://github.com/pyccel/pyccel into cuda_00

bauom committed:
added cupy support

bauom committed:
fixed type of cuda_ndaarays funcs

framdani committed:
extern "C" to function_signature added

bauom committed:
temp branch waiting for cuda_main to be merged

bauom committed:
fixed types in array_fill funcs

bauom committed:
updated CudaArray process shape

bauom committed:
removed linker options as they don't work with nvcc

bauom committed:
fixed typo

bauom committed:
updated test for array_fill

bauom committed:
added cuda tests

bauom committed:
eq in place of is

bauom committed:
capitalized first char of ccuda

bauom committed:
fixed trailing type for array_fill functions

bauom committed:
fixed trailing types

bauom committed:
specify only fortran test instead of not c or python

bauom committed:
fixed trailing type for functions in header file

bauom committed:
testing Cuda tests alone

bauom committed:
fixed trailing types

bauom committed:
test github action working dir

bauom committed:
test github action working dir

bauom committed:
removed debug and changed working dir to tests

bauom committed:
verbose True

bauom committed:
changing working dir to root

bauom committed:
removed outdated code

Pinkyboi committed:
Cuda_internal_var can only be used in kernels and device funcs

bauom committed:
reactivate normal pyccel tests

Pinkyboi committed:
Check type of passed parameter to Kernel Block number and Thread per Block number

Pinkyboi committed:
Check for stackarray variables in kernel call

framdani committed:
update clash class name for cuda

framdani committed:
activate tests

framdani committed:
fix linter

framdani committed:
define the memory location for cuda array

framdani committed:
remove trailing whitespace

framdani committed:
fix linux failing test

bauom committed:
update clash class name for cuda #1203

Update class clash name to handle name conflicts in cuda.

framdani committed:
Merge remote-tracking branch 'origin/cuda_main_temp' into cuda_array

framdani committed:
call the appropriate memory management function

framdani committed:
add missing ;

framdani committed:
update memory management functions name and memory_locations enum

Pinkyboi committed:
Add device to the check for Cuda_internal_var

Pinkyboi committed:
Check for number of passed argument to kernel call

Pinkyboi committed:
Remove check for none argument variables

Pinkyboi committed:
Revert back to previous check on passed argument

Pinkyboi committed:
Remove support for tuple and varibale number of block and number of threads per block

Pinkyboi committed:
Split check for thread per block and block number

Pinkyboi committed:
add ccuda to test languages and add the new error messages

Pinkyboi committed:
Add tests

Pinkyboi committed:
Specify test parameter inside the test file instead

Pinkyboi committed:
Make tests run only with ccuda

framdani committed:
update create_cuda_array() and add tests

bauom committed:
Merge pull request #1199 from pyccel/cuda_devel_003

Semantic stage : checks on Kernel calls

framdani committed:
Add tests for each memory_location

framdani committed:
update test for managedMemory

Pinkyboi committed:
Initial issue fix

Pinkyboi committed:
Handle case of one dimensional call

framdani committed:
remove unnecessary comments

Pinkyboi committed:
Add test for cuda_grid

Pinkyboi committed:
Call other kernels in test

framdani committed:
remove unnecessary commented lines

bauom committed:
adding multiple memory locations

adding the possibilities to use multiple memory locations: `host`, `device`, `managed`

framdani committed:
add class CudaCopy

framdani committed:
add printer and tests for cuda.copy

Pinkyboi committed:
Comment calls to func_2d func_3d

framdani committed:
add support for asynchronous copy

Pinkyboi committed:
Merge pull request #1221 from pyccel/cuda_grid

Cuda grid implementation

Pinkyboi committed:
Initial addition of memcopy_kind

Pinkyboi committed:
Raise error when allocating to host from device function

Pinkyboi committed:
Change error message for creating host array form device

Pinkyboi committed:
Change memcpy_kind variable name for readability

Pinkyboi committed:
Merge pull request #1228 from pyccel/cuda_memcopy

Addition of valid cudaMemcpyKind to cudaMemcpy

bauom committed:
Revert "Addition of valid cudaMemcpyKind to cudaMemcpy"

bauom committed:
Merge pull request #1233 from pyccel/revert-1228-cuda_memcopy

Revert "Addition of valid cudaMemcpyKind to cudaMemcpy"

bauom committed:
Merge branch 'cuda_main_temp' into cuda_copy

framdani committed:
support copy only from cuda arrays

framdani committed:
access attributes by getters

framdani committed:
add 'managed' as valid memory location arg for cuda.copy()

bauom committed:
Add cuda.copy function

Add `cuda.copy` function which copies  data from `host` or `device` and places it into a newly created array.

The function takes three parameters:
- The `src`; an existing cudaArray.
- The memory location of the `dst`; the newly created array.
- The third is a boolean which indicates whether the copy is `asynchronous` or not.

Pinkyboi committed:
Change call to cudaDeviceSynchronize to cuda.synchronize() (#1274)

bauom committed:
merged with master

bauom committed:
Temporarily deactivate most recent numpy

EmilyBourne committed:
Add coverage to cuda tests

EmilyBourne committed:
Trigger

EmilyBourne committed:
Add missing condition

EmilyBourne committed:
Add missing setup

EmilyBourne committed:
Correct name

EmilyBourne committed:
Install source

EmilyBourne committed:
Add sensible condition

EmilyBourne committed:
Install sudo

EmilyBourne committed:
Trigger for cuda_main and cuda_main_temp

EmilyBourne committed:
Don't run after failure

EmilyBourne committed:
Configure tzdata

EmilyBourne committed:
Remove ENV

EmilyBourne committed:
Require success

EmilyBourne committed:
install tzdata

EmilyBourne committed:
Try alternative definition

EmilyBourne committed:
Add check

EmilyBourne committed:
Remove debug info

EmilyBourne committed:
Put tests back on branch

EmilyBourne committed:
Fix CI tools

EmilyBourne committed:
Save xml file

EmilyBourne committed:
Save to correct xml file

EmilyBourne committed:
Don't run if failure

EmilyBourne committed:
Merge remote-tracking branch 'origin/master' into cuda_main_temp

EmilyBourne committed:
Remove restriction

Pinkyboi committed:
Fix type inference for Cupy array functions (#1300)

* Inital issue fix

* Put memory_location assignation in CupyNewArray

* Fix codacy errors

* Remove memory_location from cupy parameter / add memory_location property

* Fix the slots

* Add memory_location to infere_type for CupyNewArray

* Clean up cupy_arrange and cupy_array tests

* Make memory location a class attribute for CupyNewArray

bauom committed:
Merge branch 'master' of github.com:pyccel/pyccel into cuda_main_temp

EmilyBourne committed:
Merge remote-tracking branch 'origin/cuda_main_temp' into cuda_coverage

EmilyBourne committed:
Turn on tests for cuda

EmilyBourne committed:
Turn on tests for cuda

framdani committed:
Fix kernel launch configuration errors (#1279)

* fix kernel launch configuration errors

* fix linter

* Avoid checking for attribute 'decorators' for Interface objects

* update error messages

* add tests

* Fix new line missing

* add missing newline

* Combining error tests in a single file

* remove Trailing whitespace

* add test for invalid kernel configuration

* fix codacy issue

* trailing whitespace removed

* deactivate numpy's recent version temp

* fix typo

* Handle 'No new line at end of file'

* Fix line definition

* Add line break between errors

* Correct url

* add tests

* Activate cuda tests

* Fix cuda setup

* Update pyccel/parser/semantic.py

Co-authored-by: EmilyBourne <louise.bourne@gmail.com>

* Update variable search method

* Display correct error message

* raise error for calling non-kernel functions with parameters

* add test for invalid function call error

* codacy issues

* remove unecessary error check

* Update kernel decorator definition

* fix typo

* Fix new line missing

* Fix new line

Co-authored-by: EmilyBoune <louise.bourne@gmail.com>
Co-authored-by: bauom <mouadelalj@gmail.com>

bauom committed:
cleaning compile stage Cuda tests. (#1301)

* cleaned cuda tests to be more specific

* fixed kernel_launch function

* added kernel_launch_config tests

* fixing pyccel linter

EmilyBourne committed:
Merge remote-tracking branch 'origin/cuda_main_temp' into cuda_coverage

EmilyBourne committed:
Merge pull request #1291 from pyccel/cuda_coverage

Add coverage for cuda tests

EmilyBourne committed:
Ensure master tests are triggered on cuda_main (#1305)

Update branch with commits from master.

`cuda_main` and `cuda_devel` branches should have the same tests as
the master branch. These tests are activated here.

The setup for cuda tests is also added

Co-authored-by: Abdechahid Ihya <50792865+aihya@users.noreply.github.com>
Co-authored-by: Aymane Benaissa <47903494+Pinkyboi@users.noreply.github.com>
Co-authored-by: Youness Farini <yfarini@student.1337.ma>
Co-authored-by: Yassine Alaoui <48657685+yassine-alaoui@users.noreply.github.com>
Co-authored-by: Youness Farini <farini.work@gmail.com>
Co-authored-by: Yaman Güçlü <yaman.guclu@gmail.com>
Co-authored-by: Aaron Holmes <aaron@aaronholmes.net>
Co-authored-by: Ibrahim El Mountasser <33915853+ceciver@users.noreply.github.com>
Co-authored-by: OTHMANE HACHIM <35775290+ohachim@users.noreply.github.com>

EmilyBourne committed:
Merge remote-tracking branch 'other/master' into cuda_main

EmilyBourne committed:
Fix commit

EmilyBourne committed:
pip_installation no longer exists

EmilyBourne committed:
Add Code owners to protect merges

EmilyBourne committed:
Debug coverage problem

EmilyBourne committed:
Remove debugging

EmilyBourne committed:
Trigger codacy

EmilyBourne committed:
Trigger codacy

EmilyBourne committed:
Correct folder

EmilyBourne committed:
Include cuda files in MANIFEST

EmilyBourne committed:
Update checkout version

EmilyBourne committed:
Import coverage fixes

EmilyBourne committed:
Remove unnecessary echo

EmilyBourne committed:
Ensure cuda files are installed with pyccel (#1345)

Include cuda files in the manifest to ensure that they are installed
with pyccel. Fixes #1344 . This should help with the coverage as the
python files that are run are the installed versions not the local
versions.

Use coverage's `[paths]` options to rename files tested on different
systems. Fixes #1346 .

Pull some updates from master regarding the installation and test setup

EmilyBourne committed:
Make faster as done in cuda_main_temp

EmilyBourne committed:
Trigger codacy

EmilyBourne committed:
Try with coverage secret

EmilyBourne committed:
Add first commit to provide codacy baseline

EmilyBourne committed:
Only run benchmarks in Pyccel repo

EmilyBourne committed:
Merge remote-tracking branch 'cuda/master' into cuda_main

EmilyBourne committed:
Trigger codacy with correct branch name

EmilyBourne committed:
Merge pull request #1 from pyccel/cuda_main

Add CUDA tests to main

bauom committed:
merged pyccel-cuda/developement into cuda_main_temp

bauom committed:
complex not supported in cuda

bauom committed:
fixing cuda free

bauom committed:
fixed cuda_arange internal function

bauom committed:
fixed white spaces

bauom committed:
activate tests for developement branch

bauom committed:
typo in branch name

bauom committed:
updated ccuda codegen signature_function method

bauom committed:
merged with cuda:development

framdani committed:
Fix allocation mishandling in CUDA array creation (#1328)

This PR fixes a bug where the allocation of the CUDA array was not being
properly handled. Fixes
[pyccel-cuda#2](https://github.com/pyccel/pyccel-cuda/issues/2).

**Changes Made:**
- Fixed the allocation of the CUDA array to prevent segmentation fault.
- Added tests to ensure proper handling of the CUDA array allocation.

bauom committed:
merged pyccel:#1328

bauom committed:
remove unnecesasry comment

bauom commented on pyccel/codegen/compiling/compilers.py:
```diff
@@ -385,7 +385,7 @@ def compile_shared_library(self, compile_obj, output_folder, verbose = False, sh
         exec_cmd, includes, libs_flags, libdirs_flags, m_code = \
                 self._get_compile_components(compile_obj, accelerators)
-        linker_libdirs_flags = ['-Wl,-rpath' if l == '-L' else l for l in libdirs_flags]
+        linker_libdirs_flags = ['-Wl,-rpath' if l == '-L' and self._info['exec'] != 'nvcc' else l for l in libdirs_flags]
```
@EmilyBourne not sure if this hack is the correct way to do this but i just wanted to not use the linking flags when the compiler is `nvcc`

EmilyBourne replied:
Can you explain why you didn't want to use the linking flags?

EmilyBourne replied:
This seems to be related to this:

https://forums.developer.nvidia.com/t/solved-error-while-linking-with-nvcc/46476/6

EmilyBourne replied:
See #13 & #14

bauom commented on pyccel/codegen/pipeline.py:
```diff
@@ -198,6 +198,10 @@ def handle_error(stage):
     if compiler is None:
         compiler = 'GNU'
 
+    # Choose cuda compiler
+    if language == 'ccuda':
+        compiler = 'nvidia'
```
@EmilyBourne this is how i specify the compiler

EmilyBourne replied:
This seems very problematic as it prevents the user from providing a configuration file with different flags. The compiler should only be modified if it does not have the default value. An error could be raised if it is set to a compiler groups which we know does not support cuda , as long as you are certain that it is not likely to in future releases (but I'm not sure I know of any such groups).

bauom replied:
i think this should will be removed in some time when we will start targeting amd hardware too.
(the printers for cuda and HIP will not have a lot of differences)

EmilyBourne replied:
I think it should be removed now. It will break compilation on some systems. Instead of:
```python
    # Choose Fortran compiler
    if compiler is None:
        compiler = 'GNU'

    # Choose cuda compiler
    if language == 'ccuda':
        compiler = 'nvidia'
```
I think we should have:
```python
    # Choose default compiler family
    if compiler is None:
        compiler = 'nvidia' if language == 'ccuda' else 'GNU'
```

bauom commented on pyccel/codegen/printing/cwrappercode.py:
```diff
@@ -292,7 +292,7 @@ def static_function_signature(self, expr):
             Signature of the function.
         """
         #if target_language is C no need for the binding
-        if self._target_language == 'c':
+        if self._target_language in ('c', 'ccuda'):
```
@EmilyBourne this might need to change later if we want to bind `Ccuda` types

EmilyBourne replied:
That said the wrapper should be changed soon anyway so maybe this is ok for now

EmilyBourne replied:
> minus some specific types and function decorators.

Function decorators in cuda code?
If the types appear in the function declarations then we will indeed need a wrapper.

I imagine it also depends on how we plan to manage functions on the device. Should these be called by other translated functions or should the wrapper transfer the data to GPU to call the function?

bauom replied:
by a decorator i mean `__global__` declaration specifier.
and there is some types that are `Ccuda` specifics like complex.
for now i am always wrapping `kernel` functions inside another simple `C` function because we do not have a proper wrapper.
I do not think there will be any problem with the function on the `device` because all of them can only interact with the `kernel`.

bauom committed:
removed debug print

bauom committed:
fixed alignement

bauom commented on pyccel/decorators.py:
```diff
@@ -98,3 +100,7 @@ def allow_negative_index(f,*args):
         return f
     return identity
+
+def kernel(f):
+    return numpy.array([[f]])
+
```
@EmilyBourne not sure if we can make this available only from the `cuda` module.

EmilyBourne replied:
So instead of doing `from pyccel.decorators import kernel` we would do `from pyccel.cuda.decorators import kernel` or (with a well-defined `__init__.py`) `from pyccel.cuda import kernel`?
This seems ok.

I'm not sure I understand your question "not sure if we can". Do you mean "can" or "should"?

bauom replied:
when i was writing that code i didn't find a way to do this or maybe i just didn't think enough :sweat_smile:

bauom commented on pyccel/naming/ccudanameclashchecker.py:
```diff
@@ -0,0 +1,69 @@
+class CCudaNameClashChecker(metaclass = Singleton):
+    """ Class containing functions to help avoid problematic names in Ccuda
+    """
+    # Keywords as mentioned on https://en.cppreference.com/w/c/keyword
```
we still need to add some of these : https://icl.utk.edu/~mgates3/docs/cuda.html

bauom commented on pyccel/parser/semantic.py:
```diff
@@ -510,6 +513,28 @@ def _infer_type(self, expr, **settings):
             d_var['cls_base'   ] = NumpyArrayClass
             return d_var
 
+        elif isinstance(expr, CupyNewArray):
+            d_var['datatype'   ] = expr.dtype
+            d_var['memory_handling'] = 'heap' if expr.rank > 0 else 'stack'
+            d_var['memory_location'] = expr.memory_location
+            d_var['shape'      ] = expr.shape
+            d_var['rank'       ] = expr.rank
+            d_var['order'      ] = expr.order
+            d_var['precision'  ] = expr.precision
+            d_var['cls_base'   ] = CudaArrayClass
+            return d_var
+
+        elif isinstance(expr, CudaNewArray):
+            d_var['datatype'   ] = expr.dtype
+            d_var['memory_handling'] = 'heap' if expr.rank > 0 else 'stack'
+            d_var['memory_location'] = expr.memory_location
+            d_var['shape'      ] = expr.shape
+            d_var['rank'       ] = expr.rank
+            d_var['order'      ] = expr.order
+            d_var['precision'  ] = expr.precision
+            d_var['cls_base'   ] = CudaArrayClass
+            return d_var
```
@EmilyBourne
- not sure if we should have two of these
- not sure why we have two of these

EmilyBourne replied:
Ah no they're not the same. It's `CudaNewArray` and `CupyNewArray`. Do we need both these classes though? I still haven't understood the difference if we do then it should be `elif isinstance(expr, (CudaNewArray, CupyNewArray))`

bauom replied:
yeah it should just be one of them as `Cupy` arrays should inherit `CudaNewArray`

EmilyBourne replied:
Ok. This just moves the question though. What is the difference between `CudaArray` and `CupyArray`?

bauom replied:
`CudaArray` is an internal function that uses `Ccuda` API for allocations.
it has more freedom to allocate to the `device`/`host`/`managed` memory locations.

`CupyArray` follows the Cupy and only allocate to the `device`.

EmilyBourne replied:
Maybe I'm just being stupid but I still don't see a need for both of these types. Isn't a `CupyArray` equivalent to a `CudaArray(memory_location='device')`?

When do you use `CudaArray` rather than `CupyArray`? Is a `CupyArray` converted to a `CudaArray` for printing? What Python code generates a `CudaArray`?

bauom replied:
I agree it needs cleanup.

bauom commented on pyccel/parser/semantic.py:
```diff
@@ -814,8 +839,17 @@ def _handle_function(self, expr, func, args, **settings):
         =======
         new_expr : FunctionCall or PyccelInternalFunction
         """
+        if isinstance(func, FunctionDef) and 'kernel' in func.decorators:
+            errors.report(MISSING_KERNEL_CONFIGURATION, symbol = expr, severity = 'fatal')
```
@EmilyBourne  not sure if this is the correct place of raising these errors

EmilyBourne replied:
If this is where you detect it then it seems fine to me. Where else would you want to raise it?

bauom commented on pyccel/parser/semantic.py:
```diff
@@ -814,8 +839,17 @@ def _handle_function(self, expr, func, args, **settings):
+            errors.report(MISSING_KERNEL_CONFIGURATION, symbol = expr, severity = 'fatal')
         if isinstance(func, PyccelFunctionDef):
             func = func.cls_name
+            if func in (CudaThreadIdx, CudaBlockDim, CudaBlockIdx, CudaGridDim):
+                if 'kernel' not in self.scope.decorators\
+                    and 'device' not in self.scope.decorators:
+                    errors.report("Cuda internal variables should only be used in Kernel or Device functions",
+                        symbol = expr,
+                        severity = 'fatal')
```
@EmilyBourne not sure if this is the correct place of raising these errors

EmilyBourne replied:
If this is where you detect it then it seems fine to me. Where else would you want to raise it?

bauom commented on pyccel/stdlib/ndarrays/ndarrays.c:
```diff
@@ -42,12 +42,14 @@ t_ndarray   array_create(int32_t nd, int64_t *shape,
         case nd_bool:
             arr.type_size = sizeof(bool);
             break;
+        #ifndef __NVCC__
         case nd_cfloat:
             arr.type_size = sizeof(float complex);
             break;
         case nd_cdouble:
             arr.type_size = sizeof(double complex);
             break;
+        #endif
```
@EmilyBourne Complex are not supported for now as we should handle them in a special way

bauom replied:
the macro is set to true when compiling with the nvidia cuda compiler `nvcc` where we do not have the support of `float complex` and `double complex` and we should use `#include <cuComplex.h>` instead when compiling with nvcc.

bauom replied:
`nvcc` is focused on CUDA programming and is compatible with all NVIDIA GPU architectures.
`nvc` is designed for HPC applications and is optimized for the latest NVIDIA GPU architectures.

EmilyBourne replied:
According to [Nvidia's docs](https://docs.nvidia.com/hpc-sdk/index.html), `nvcc` is a CUDA compiler while `nvc` is a C11 compiler

EmilyBourne replied:
So `#ifndef __NVCC__` sounds like the right thing, but I'm still not clear why we can't do:
```c
#ifndef __NVCC__
#include <complex>
#else
#include <cuComplex.h>
#endif
```

bauom replied:
the `cuComplex` are not compatible with the C ones i think we need to do what you suggested but need to add the wrapping.
and they use their own API for the operations.
will be added in a separate issue.

bauom reviewed (commented):
@EmilyBourne other than the mentions 
I would like you to take a look on how we implimented the `Kernel` class and how we handled that.
we hacked it in some places like syntactic but i am not sure if that is the only way to do it.

bauom committed:
fix pytest cuda action

EmilyBourne commented:
The lint error needs fixing (`_infere_type`->`_infer_type` was renamed to remove the typo), but from a coverage perspective it is worrying if this difference is not leading to errors

EmilyBourne commented on pyccel/ast/core.py:
```diff
@@ -74,10 +74,11 @@
     'Import',
-    'InlineFunctionDef',
     'InProgram',
+    'InlineFunctionDef',
```
Any idea why this was moved? The others seem to be alphabetical, but not here

bauom replied:
all the moves should be alphabetical as far as i know i think the `P` < `l` in `Program` and `line`

EmilyBourne replied:
Ah! That explains it. It's not alphabetical but it is ordered by characters:+1:

EmilyBourne committed:
Fix triggers

EmilyBourne committed:
Merge branch 'development' into cuda_main_temp

EmilyBourne commented on pyccel/ast/core.py:
```diff
@@ -2150,6 +2151,60 @@ def prefix(self):
+        self._tpblock = tpblock
+        super().__init__(func, args, current_function)
+
+    @property
+    def func(self):
+        """ The number of blocks in which the kernel will run
+        """
+        return self._func
```
Why is this property needed as well as the inherited `funcdef` property?

bauom replied:
fixed in [0a8272c](https://github.com/pyccel/pyccel-cuda/pull/12/commits/0a8272c4f666d4d4b5a2b920daf7bb440400e205)

EmilyBourne commented on pyccel/ast/core.py:
```diff
@@ -2150,6 +2151,60 @@ def prefix(self):
+
+    def __init__(self, func, args, numBlocks, tpblock, current_function=None):
+
+        self._func = func
+        self._args = args
```
I'm not sure these are needed. Aren't they handled in the super class?

EmilyBourne commented:
> @EmilyBourne other than the mentions I would like you to take a look on how we implimented the `Kernel` class and how we handled that. we hacked it in some places like syntactic but i am not sure if that is the only way to do it.

I can't find the `Kernal` class. Do you mean the `KernalCall` class?

EmilyBourne commented on pyccel/ast/cudaext.py:
```diff
@@ -0,0 +1,309 @@
+    _attribute_nodes = ()
+    def __init__(self):
+        #...
+        self._shape     = None
+        self._rank      = 0
+        self._dtype     = NativeInteger()
+        self._precision = None
+        self._order     = None
```
If these are the same for all instances of the class then they should be class variables not object variables

EmilyBourne commented on pyccel/ast/cudaext.py:
```diff
@@ -0,0 +1,309 @@
+            raise ValueError("dimension need to be 0, 1 or 2")
+        #...
+        self._dim       = dim
+        self._shape     = None
+        self._rank      = 0
```
See above, class variables

bauom replied:
changed in [09b3533](https://github.com/pyccel/pyccel-cuda/pull/12/commits/09b35337fe4c66a4a28bacd621adef5d31d31def)

EmilyBourne commented on pyccel/ast/cudaext.py:
```diff
@@ -0,0 +1,309 @@
+        
+        if isinstance(dim, int):
+            dim = LiteralInteger(dim)
+        if not isinstance(dim, LiteralInteger):
+            raise TypeError("dimension need to be an integer")
```
I am a little confused by this. According to the docstring this is an internal variable. I am not sure exactly what is meant by this but I assume it is not created by the user. In that case why do we need to use `LiteralInteger` rather than integer?

bauom replied:
this is just a general class for the internal variables.
which are called in the code as `CudaThreadIdx(x)` where x can be `0, 1, 2`.
i think a check for `PyccelAstNode` would be better.

Note: this will change later as we discussed to using class_def instead. maybe another Pr?

EmilyBourne replied:
I agree that this could be done in another PR, but in that case we need an issue opening which explains the fix clearly

EmilyBourne replied:
That said it may be easier to merge that PR into here to avoid having to write the doc strings for all these unnecessary objects ;P

EmilyBourne replied:
See #16

EmilyBourne commented on pyccel/ast/cudaext.py:
```diff
@@ -0,0 +1,309 @@
+        self._order     = None
+        super().__init__()
+
+class CudaInternalVar(PyccelAstNode):
+    """
+    Represents a General Class For Cuda internal Variables Used To locate Thread In the GPU architecture"
```
I don't understand the purpose of this class. Is `CudaInternalVar(LiteralInteger(2))` equivalent to `Variable('int','X', is_temp=True)`? Or is it not printed? If it isn't printed then is a class needed for this? It only seems to store `0`/`1`/`2` and doesn't do anything more complicated

EmilyBourne commented on pyccel/ast/cudaext.py:
```diff
@@ -0,0 +1,309 @@
+    def is_async(self):
+        return self._is_async
+
+class CudaThreadIdx(CudaInternalVar):
+    __slots__ = ()
+    pass
+class CudaBlockDim(CudaInternalVar):
+    __slots__ = ()
+    pass
+class CudaBlockIdx(CudaInternalVar):
+    __slots__ = ()
+    pass
+class CudaGridDim(CudaInternalVar):
+    __slots__ = ()
+    pass
```
`pass` is not needed if `__slots__ = ()` is provided. See discussion above. Should these be classes or instances of `Variable`?

EmilyBourne commented on pyccel/ast/cudaext.py:
```diff
@@ -0,0 +1,309 @@
+
+class CudaGrid(PyccelAstNode)               :
+    """
+    CudaGrid locate Thread In the GPU architecture Using CudaThreadIdx, CudaBlockDim, CudaBlockIdx
+    To calculate the exact index of the thread automatically.
```
Some of these docstrings have capitalisation in the middle of the sentence. This needs fixing it is not so clear to read. In English capital letters are used for the first word in a sentence, the name of something, or abbreviations.

```suggestion
    CudaGrid locates a thread in the GPU architecture using `CudaThreadIdx`, `CudaBlockDim`, and `CudaBlockIdx`
    to calculate the exact index of the thread automatically.
```

EmilyBourne commented on pyccel/ast/cudaext.py:
```diff
@@ -0,0 +1,309 @@
+    'CudaGridDim'   : 'gridDim'
+}
+
+cuda_constants = {
+
+}
```
Is this necessary?

bauom replied:
as i was passing it to the `variables` attr of the class `Module` i made it like that but not i am just passing `[]` so i have removed it.
[9f989ff](https://github.com/pyccel/pyccel-cuda/pull/12/commits/9f989ff8e53005632eef4f421609c3d8ed4db490)

EmilyBourne commented on pyccel/ast/utilities.py:
```diff
@@ -78,12 +80,15 @@ def builtin_function(expr, args=None):
 pyccel_mod = Module('pyccel',(),(),
-        imports = [Import('decorators', decorators_mod)])
+        imports = [Import('decorators', decorators_mod),
+                    Import('cuda', AsName(cuda_mod,'cuda')),
```
I don't think the `AsName` is needed here

bauom replied:
yes it is not needed.
is there a difference between this and the imports in `builtin_import_registery`?
[here](https://github.com/pyccel/pyccel-cuda/blob/ba35a691d261c91d34e464715be20c03a13d7117/pyccel/ast/utilities.py#L88)

bauom replied:
[c7222ff](https://github.com/pyccel/pyccel-cuda/pull/12/commits/c7222ff4882df1901dffbd1b1d82caa7a4eef674)

EmilyBourne replied:
Ah! This is my fault :facepalm: . The second argument to an `Import` has to be either a `Module` or an `AsName`. I obviously forgot when I was writing `builtin_import_registery`

EmilyBourne replied:
Fixed https://github.com/pyccel/pyccel/pull/1364

EmilyBourne commented on pyccel/codegen/printing/ccode.py:
```diff
@@ -412,12 +412,14 @@ def arrayFill(self, expr):
         declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
-
+        dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
+        dtype = dtype[3:]
```
Why do we cut the first 3 characters? Should we have a type_registry instead of a `ndarray_type_registry`?

bauom replied:
yeah maybe a type registry will be better and we add the `nd_` prefix manually.

EmilyBourne replied:
I think this can be a separate issue/PR

EmilyBourne replied:
https://github.com/pyccel/pyccel-cuda/issues/17

EmilyBourne commented on pyccel/codegen/printing/ccode.py:
```diff
@@ -412,12 +412,14 @@ def arrayFill(self, expr):
+                code_init += 'array_fill_{0}(({1}){2}, {3});\n'.format(dtype, declare_dtype, self._print(rhs.fill_value), self._print(lhs))
             else:
-                code_init += 'array_fill({0}, {1});\n'.format(self._print(rhs.fill_value), self._print(lhs))
+                code_init += 'array_fill_{0}({1}, {2});\n'.format(dtype, self._print(rhs.fill_value), self._print(lhs))
```
Is there a difference between `declare_dtype` and `dtype`?

bauom replied:
`declare_dtype`: have the type name as written in C, we use them in the declaration.
`dtype` : has the name used in `ndarrays` funcs to pass the types and in the union to define the type of the array. they are written as `nd_`+`abv of the type`.

EmilyBourne commented on pyccel/codegen/printing/cwrappercode.py:
```diff
@@ -360,6 +360,7 @@ def get_static_declare_type(self, variable):
         str
             The code describing the type.
         """
+
```
Useless change

EmilyBourne commented on pyccel/commands/console.py:
```diff
@@ -60,7 +60,7 @@ def pyccel(files=None, mpi=None, openmp=None, openacc=None, output_dir=None, com
     group = parser.add_argument_group('Backend compiler options')
 
-    group.add_argument('--language', choices=('fortran', 'c', 'python'), help='Generated language')
+    group.add_argument('--language', choices=('fortran', 'c', 'python', 'ccuda'), help='Generated language')
```
There should be an equivalent change in the `epyccel` docstring

bauom replied:
[8e318c5](https://github.com/pyccel/pyccel-cuda/pull/12/commits/8e318c5644c41aaa4e7d721345179bc5fcf3d297)

EmilyBourne commented on pyccel/decorators.py:
```diff
@@ -6,6 +6,7 @@
 """
 This module contains all the provided decorator methods.
 """
+import numpy
```
I would prefer just importing what we need:
```suggestion
from numpy import array
```

EmilyBourne commented on pyccel/decorators.py:
```diff
@@ -98,3 +100,7 @@ def allow_negative_index(f,*args):
     return identity
+
+def kernel(f):
+    return numpy.array([[f]])
```
I think we should add a comment explaining that this is written like this to allow the function to be indexed and the python code to execute without errors

EmilyBourne commented on pyccel/errors/messages.py:
```diff
@@ -160,3 +160,10 @@
 WRONG_LINSPACE_ENDPOINT = 'endpoint argument must be boolean'
 NON_LITERAL_KEEP_DIMS = 'keep_dims argument must be a literal, otherwise rank is unknown'
 NON_LITERAL_AXIS = 'axis argument must be a literal, otherwise pyccel cannot determine which dimension to operate on'
+KERNEL_STACK_ARRAY_ARG = "A variable allocated on the stack can't be passed to a Kernel function"
```
Does this cause compiler errors?

EmilyBourne commented on pyccel/parser/syntactic.py:
```diff
@@ -896,6 +896,10 @@ def _visit_Call(self, stmt):
             func = DottedName(*func.name[:-1], func_attr)
+        elif isinstance(func, IndexedElement):
+            if len(func.indices) != 2:
+                raise NotImplementedError
```
Missing error message

EmilyBourne commented on pyccel/parser/syntactic.py:
```diff
@@ -896,6 +896,10 @@ def _visit_Call(self, stmt):
         elif isinstance(func, DottedName):
             func_attr = FunctionCall(func.name[-1], args)
             func = DottedName(*func.name[:-1], func_attr)
+        elif isinstance(func, IndexedElement):
+            if len(func.indices) != 2:
+                raise NotImplementedError
+            func = KernelCall(func.base, args, func.indices[0], func.indices[1])
         else:
```
This looks ok to me

EmilyBourne commented on pyccel/naming/ccudanameclashchecker.py:
```diff
@@ -0,0 +1,69 @@
+        'cuda_array_fill_int32', 'cuda_array_fill_int8',
+        'cuda_array_arange_double', 'cuda_array_arange_int64', 
+        'cuda_array_arange_int32', 'cuda_array_arange_int8',
+        'cudaMallocManaged', 'cudaSynchronize'])
```
I wonder if this should extend `CNameClashChecker.keywords`?

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -119,7 +123,7 @@
 from pyccel.errors.errors import PyccelSemanticError
 
-from pyccel.errors.messages import (PYCCEL_RESTRICTION_TODO, UNDERSCORE_NOT_A_THROWAWAY,
+from pyccel.errors.messages import (INVALID_FUNCTION_CALL, INVALID_KERNEL_CALL_BP_GRID, INVALID_KERNEL_CALL_TP_BLOCK, MISSING_KERNEL_CONFIGURATION,PYCCEL_RESTRICTION_TODO, UNDERSCORE_NOT_A_THROWAWAY,
```
Line too long

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -870,6 +904,92 @@ def _handle_function(self, expr, func, args, **settings):
 
+    def _handle_kernel(self, expr, func, args, **settings):
+        """
+        Create a FunctionCall or an instance of a PyccelInternalFunction
+        from the function information and arguments
```
Is this description accurate or copy/pasted from `_handle_function`?

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -870,6 +904,92 @@ def _handle_function(self, expr, func, args, **settings):
+        ==========
+        expr : PyccelAstNode
+               The expression where this call is found (used for error output)
+        func : FunctionDef instance, Interface instance or PyccelInternalFunction type
+               The function being called
```
Wrong type

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -870,6 +904,92 @@ def _handle_function(self, expr, func, args, **settings):
+            args, kwargs = split_positional_keyword_arguments(*args)
+            for a in args:
+                if getattr(a,'dtype',None) == 'tuple':
+                    self._infere_type(a, **settings)
+            for a in kwargs.values():
+                if getattr(a,'dtype',None) == 'tuple':
+                    self._infere_type(a, **settings)
```
Using old function name:

```suggestion
                    self._infer_type(a, **settings)
            for a in kwargs.values():
                if getattr(a,'dtype',None) == 'tuple':
                    self._infer_type(a, **settings)
```

EmilyBourne replied:
This code is untested

EmilyBourne replied:
This code isn't reported as untested but I'm not sure how that can be true if it's using a function that doesn't exist. Something must be behaving weirdly

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -870,6 +904,92 @@ def _handle_function(self, expr, func, args, **settings):
+                        symbol = expr,
+                        severity='fatal')
+            # TODO : type check the NUMBER OF BLOCKS 'numBlocks' and threads per block 'tpblock'
+            if not isinstance(expr.numBlocks, LiteralInteger):
+                # expr.numBlocks could be invalid type, or PyccelSymbol
+                if isinstance(expr.numBlocks, PyccelSymbol):
+                    numBlocks = self.get_variable(expr.numBlocks)
+                    if not isinstance(numBlocks.dtype, NativeInteger):
+                        errors.report(INVALID_KERNEL_CALL_BP_GRID,
+                        symbol = expr,
+                        severity='error')
+                else:
+                    errors.report(INVALID_KERNEL_CALL_BP_GRID,
+                        symbol = expr,
+                        severity='error')
```
I find this quite hard to follow. I suggest:
```suggestion
                if isinstance(expr.numBlocks, PyccelSymbol):
                    numBlocks = self.get_variable(expr.numBlocks)
                if not isinstance(numBlocks.dtype, NativeInteger):
                    errors.report(INVALID_KERNEL_CALL_BP_GRID,
                        symbol = expr,
                        severity='error')
```

EmilyBourne replied:
Hmm, you're right. I didn't look closely enough and assumed it was mentioned at the start. Accessing `expr.numBlocks` (via a property) is more costly than accessing a variable directly so usually we prefer defining a variable once rather than using a property multiple times (I need to write this in a doc) . I missed that this wasn't the case here.
```suggestion
            numBlocks = expr.numBlocks
            if isinstance(numBlocks, PyccelSymbol):
                    numBlocks = self.get_variable(numBlocks)
            if not isinstance(numBlocks.dtype, NativeInteger):
                    errors.report(INVALID_KERNEL_CALL_BP_GRID,
                        symbol = expr,
                        severity='error')
```

EmilyBourne replied:
Doc written. See https://github.com/pyccel/pyccel/pull/1357

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -870,6 +904,92 @@ def _handle_function(self, expr, func, args, **settings):
+                else:
+                    errors.report(INVALID_KERNEL_CALL_TP_BLOCK,
+                        symbol = expr,
+                        severity='error')
```
See above. Repeated error and hard to follow so many ifs

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -870,6 +904,92 @@ def _handle_function(self, expr, func, args, **settings):
+                        symbol = expr,
+                        severity='error')
+                elif isinstance(a.value, Variable) and a.value.on_stack:
+                    errors.report("A variable allocated on the stack can't be passed to a Kernel function",
```
Why not use the identical error from errors.messages?

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -870,6 +904,92 @@ def _handle_function(self, expr, func, args, **settings):
+            if isinstance(func, FunctionDef):
+                self._check_argument_compatibility(new_expr.args, func.arguments,
+                        expr, func.is_elemental)
+            return new_expr
```
This seems to duplicate lots of code from `_handle_function` is there no way to collect all the differences together and call `_handle_function` directly?

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -1856,7 +1976,6 @@ def _visit_PyccelSymbol(self, expr, **settings):
 
 
     def _visit_DottedName(self, expr, **settings):
-
```
Unnecessary change

EmilyBourne commented on pyccel/parser/semantic.py:
```diff
@@ -2011,6 +2130,19 @@ def _visit_DottedName(self, expr, **settings):
+        args = self._handle_function_args(expr.args, **settings)
+
+        return self._handle_kernel(expr, func, args, **settings)
+        # return KernelCall(func, expr.args, expr.numBlocks, expr.tpblock)
```
Remove dead code

EmilyBourne commented on pyccel/stdlib/cuda_ndarrays/cuda_ndarrays.cu:
```diff
@@ -0,0 +1,170 @@
+        arr.nd_int8[i] = (i + start);
+}
+__global__
+void cuda_array_arange_int32(t_ndarray arr, int start)
```
I find it odd that you have `int8` and `int32` but not `int16`. Is this intentional?

EmilyBourne commented on pyccel/stdlib/cuda_ndarrays/cuda_ndarrays.cu:
```diff
@@ -0,0 +1,170 @@
+        arr.nd_double[i] = c;
+}
+
+void    device_memory(void** devPtr, size_t size)
+{
+    cudaMalloc(devPtr, size);
+}
+
+void    managed_memory(void** devPtr, size_t size)
+{
+    cudaMallocManaged(devPtr, size);
+}
+
+void    host_memory(void** devPtr, size_t size)
+{
+    cudaMallocHost(devPtr, size);
+}
```
I find the names of these functions misleading. It is not clear that `device_memory` allocates memory rather than just accessing it

EmilyBourne replied:
Sounds good to me :+1:

EmilyBourne commented on pyccel/stdlib/ndarrays/ndarrays.c:
```diff
@@ -168,7 +172,7 @@ void   _array_fill_float(float c, t_ndarray arr)
 }
 
-void   _array_fill_double(double c, t_ndarray arr)
+void   array_fill_double(double c, t_ndarray arr)
```
Why the name change? Should this be merged to pyccel/pyccel?

EmilyBourne commented:
As a general remark, why do we use "ccuda" everywhere instead of just "cuda"? The language is cuda. It is c-based, but I don't think there is a "fortrancuda" so I'm not sure where this naming has come from

EmilyBourne commented on pyccel/commands/console.py:
```diff
@@ -60,7 +60,7 @@ def pyccel(files=None, mpi=None, openmp=None, openacc=None, output_dir=None, com
     group = parser.add_argument_group('Backend compiler options')
 
-    group.add_argument('--language', choices=('fortran', 'c', 'python'), help='Generated language')
+    group.add_argument('--language', choices=('fortran', 'c', 'python', 'ccuda'), help='Generated language')
```
Why `--language ccuda` rather than `--language cuda` ?

EmilyBourne commented on tests/ndarrays/test_ndarrays.c:
```diff
@@ -609,7 +609,7 @@ int32_t test_array_fill_int64(void)
 
     x = array_create(2, m_1_shape, nd_int64, false);
-    array_fill((int64_t)32, x);
+    array_fill_int64((int64_t)32, x);
```
Is this name explicitly because the macro was removed with the nvcc checks?

EmilyBourne commented on tests/internal/scripts/ccuda/kernel_launch_config_variable.py:
```diff
@@ -0,0 +1,13 @@
+from pyccel.decorators import kernel
```
Please add pylint comments to test files for documentation. This applies to lots of files

EmilyBourne commented on tests/cuda_test/test_kernel_semantic.py:
```diff
@@ -0,0 +1,218 @@
+    # Check that the error is correct
+    error_info = [*errors.error_info_map.values()][0][0]
+    assert error_info.symbol.func  == 'stack_array_kernel'
+    assert KERNEL_STACK_ARRAY_ARG == error_info.message
```
Could this test ever be run with a different language? If not then I don't see any reason to pass this as an argument

EmilyBourne commented on samples/test/test.py:
```diff
@@ -0,0 +1,28 @@
+    i = cuda.threadIdx(0)
+    """
+    test test
+    """
```
Is this supposed to test a docstring? If so then it is in the wrong place

EmilyBourne commented on samples/test/tt.cu:
```diff
@@ -0,0 +1,20 @@
+    cudaDeviceSynchronize();
+    printf("%d\n", p1[0]);
+    return 0;
+}
```
What is the `samples` folder? <was it committed by accident or should it be mentioned in docs?

EmilyBourne commented on pyccel/ast/cupyext.py:
```diff
@@ -0,0 +1,488 @@
+        self._shape = shape
+        self._rank  = len(self._shape)
+        self._dtype = dtype
+        self._order = CupyNewArray._process_order(self._rank, order)
```
Does `CupyNewArray` need to be named explicitly here? I'm not sure of best practice here. Maybe this is ok, or maybe we should prefer `self` for ease in case of renaming/reorganisation of the inheritance

EmilyBourne commented on pyccel/ast/cupyext.py:
```diff
@@ -0,0 +1,488 @@
+    @property
+    def fill_value(self):
+        dtype = self.dtype
+        if isinstance(dtype, NativeInteger):
+            value = LiteralInteger(0, precision = self.precision)
+        elif isinstance(dtype, NativeFloat):
+            value = LiteralFloat(0, precision = self.precision)
+        elif isinstance(dtype, NativeComplex):
+            value = LiteralComplex(0., 0., precision = self.precision)
+        elif isinstance(dtype, NativeBool):
+            value = LiteralFalse(precision = self.precision)
+        else:
+            raise TypeError('Unknown type')
```
Why not use `convert_to_literal` ?

```suggestion
        value = convert_to_literal(0, dtype, self.precision)
```

EmilyBourne commented on pyccel/ast/cupyext.py:
```diff
@@ -0,0 +1,488 @@
+    @property
+    def fill_value(self):
+        dtype = self.dtype
+        if isinstance(dtype, NativeInteger):
+            value = LiteralInteger(1, precision = self.precision)
+        elif isinstance(dtype, NativeFloat):
+            value = LiteralFloat(1., precision = self.precision)
+        elif isinstance(dtype, NativeComplex):
+            value = LiteralComplex(1., 0., precision = self.precision)
+        elif isinstance(dtype, NativeBool):
+            value = LiteralTrue(precision = self.precision)
+        else:
+            raise TypeError('Unknown type')
```
See above, `convert_to_literal`

EmilyBourne commented on pyccel/ast/cupyext.py:
```diff
@@ -0,0 +1,488 @@
+        elif all(isinstance(s, LiteralInteger) for s in a.shape):
+            return LiteralInteger(reduce(operator.mul, [s.python_value for s in a.shape]))
+        else:
+            return super().__new__(cls)
```
What is the difference between this class and `NumpyArraySize`? Just the name?

EmilyBourne commented on pyccel/ast/cupyext.py:
```diff
@@ -0,0 +1,488 @@
+
+cuda_constants = {
+}
+
```
Is `cuda_constants` needed if there are no constants?

EmilyBourne commented on pyccel/codegen/printing/ccudacode.py:
```diff
@@ -0,0 +1,591 @@
+            str
+                Signature of the function.
+            """
+            if len(expr.results) > 1:
+                self._additional_args.append(expr.results)
+            args = list(expr.arguments)
+            if len(expr.results) == 1:
+                ret_type = self.get_declare_type(expr.results[0])
+            elif len(expr.results) > 1:
+                ret_type = self._print(datatype('int'))
+                args += [FunctionDefArgument(a) for a in expr.results]
+            else:
+                ret_type = self._print(datatype('void'))
+            name = expr.name
+            if not args:
+                arg_code = 'void'
+            else:
+                def get_var_arg(arg, var):
+                    code = "const " * var.is_const
+                    code += self.get_declare_type(var) + ' '
+                    code += arg.name * print_arg_names
+                    return code
+
+                var_list = [a.var for a in args]
+                arg_code_list = [self.function_signature(var, False) if isinstance(var, FunctionAddress)
+                                    else get_var_arg(arg, var) for arg, var in zip(args, var_list)]
+                arg_code = ', '.join(arg_code_list)
+
+            if self._additional_args :
+                self._additional_args.pop()
+
+            extern_word = 'extern "C"'
+            cuda_deco = "__global__" if 'kernel' in expr.decorators else ''
+
+            if isinstance(expr, FunctionAddress):
+                return f'{extern_word} {ret_type} (*{name})({arg_code})'
+            else:
+                return f'{extern_word} {cuda_deco} {ret_type} {name}({arg_code})'
```
A lot of this seems similar to `CCodePrinter.function_signature`. Can we not use the inheritance?

```suggestion
            extern_word = 'extern "C"'
            cuda_deco = "__global__" if 'kernel' in expr.decorators else ''
            c_signature = CCodePrinter.function_signature(self, expr, print_arg_names)

            if isinstance(expr, FunctionAddress):
                return f'{extern_word} {c_signature}'
            else:
                return f'{extern_word} {cuda_deco} {c_signature}'
```

EmilyBourne commented on pyccel/codegen/printing/ccudacode.py:
```diff
@@ -0,0 +1,591 @@
+        free_code = ''
+        #free the array if its already allocated and checking if its not null if the status is unknown
+        if  (expr.status == 'unknown'):
+            free_code = 'if (%s.shape != NULL)\n' % self._print(expr.variable.name)
+            free_code += "{{\n{}}}\n".format(self._print(Deallocate(expr.variable)))
+        elif  (expr.status == 'allocated'):
+            free_code += self._print(Deallocate(expr.variable))
+        shape = ", ".join(self._print(i) for i in expr.shape)
+        shape_dtype = self.find_in_dtype_registry('int', 8)
+        tmp_shape = self.scope.get_new_name('tmp_shape')
+        dtype = self._print(expr.variable.dtype)
```
The `expr.variable` property is used a lot here. We should declare `variable` (see https://github.com/pyccel/pyccel/pull/1357)

EmilyBourne commented on pyccel/codegen/printing/ccudacode.py:
```diff
@@ -0,0 +1,591 @@
+        dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
+        dtype = dtype[3:]
+
+        #TODO: calculate best thread number to run the kernel
```
Is this TODO up to date? Is there an issue for it?

EmilyBourne commented on pyccel/codegen/printing/ccudacode.py:
```diff
@@ -0,0 +1,591 @@
+        rhs = expr.rhs
+        lhs = expr.lhs
+        if rhs.rank == 0:
+            raise NotImplementedError(str(expr))
```
I am not convinced that this error is sufficiently explicit

EmilyBourne commented on pyccel/codegen/printing/ccudacode.py:
```diff
@@ -0,0 +1,591 @@
+        self.add_import(c_imports['string'])
+        if isinstance(arg, Variable):
+            arg = self._print(arg)
+            cpy_data = "cudaMemcpy({0}.raw_data, {1}.{2}, {0}.buffer_size, cudaMemcpyHostToDevice);".format(lhs, arg, dtype)
```
We shouldn't use `.` explicitly, rather we should use `DottedName` etc to ensure we have the neatest printing possible. See https://github.com/pyccel/pyccel/blob/master/developer_docs/codegen_stage.md#use-of-c-pointers

EmilyBourne commented on pyccel/codegen/printing/ccudacode.py:
```diff
@@ -0,0 +1,591 @@
+        else :
+            arg = ', '.join(self._print(i) for i in arg)
+            dummy_array = "%s %s[] = {%s};\n" % (declare_dtype, dummy_array_name, arg)
+            cpy_data = "cudaMemcpy({0}.raw_data, {1}, {0}.buffer_size, cudaMemcpyHostToDevice);".format(self._print(lhs), dummy_array_name, dtype)
```
See above https://github.com/pyccel/pyccel/blob/master/developer_docs/codegen_stage.md#use-of-c-pointers

EmilyBourne commented on pyccel/codegen/printing/ccudacode.py:
```diff
@@ -0,0 +1,591 @@
+            to_location = 'Device'
+        transfer_type = 'cudaMemcpy{0}To{1}'.format(from_location, to_location)
+        if isinstance(rhs.is_async, LiteralTrue):
+            cpy_data = "cudaMemcpyAsync({0}.raw_data, {1}.raw_data, {0}.buffer_size, {2}, 0);".format(lhs, rhs.arg, transfer_type)
```
See above https://github.com/pyccel/pyccel/blob/master/developer_docs/codegen_stage.md#use-of-c-pointers

EmilyBourne committed:
Merge branch 'pyccel:master' into master

EmilyBourne committed:
Merge branch 'master' into cuda_master

EmilyBourne committed:
Set up bot for cuda

EmilyBourne committed:
Correct double if

EmilyBourne committed:
Wrong file

EmilyBourne committed:
Merge branch 'pyccel:master' into master

EmilyBourne committed:
Merge remote-tracking branch 'origin/master' into cuda_master

EmilyBourne commented:
/bot trust user bauom

EmilyBourne commented:
/bot trust user framdani

EmilyBourne committed:
Correct typo

EmilyBourne commented:
/bot trust user bauom

EmilyBourne commented:
/bot trust user framdani

EmilyBourne committed:
Use rpath with nvidia (#14)

This PR adds the -Xcompiler flag to the CUDA code to pass the -Wl,-rpath flags to the linker. The -Wl,-rpath flags are required for the linker to find shared libraries that are not on the path. By using the -Xcompiler flag, the flags are passed to the linker as expected, ensuring that the executable can find the required shared libraries. This should fix the issue where the executable fails to find shared libraries without the -Wl,-rpath flags.

bauom committed:
Merge branch 'development' into cuda_main_temp

bauom commented:
/bot run linux, cuda

bauom commented:
/bot show tests

EmilyBourne commented:
@bauom the bot looks for test names separated by spaces `linux,` will be unknown

bauom commented:
/bot run linux cuda

EmilyBourne committed:
Missing bot output

bauom committed:
Merge branch 'development' into cuda_main_temp

EmilyBourne committed:
Merge branch 'pyccel:master' into master

bauom committed:
Merge branch 'development' into cuda_main_temp

bauom committed:
changed how we choose default compiler family

bauom committed:
removed dead code

bauom committed:
removed unnecessary property func

bauom committed:
added missed change to tests

bauom committed:
remove unnecessary code in kernelcall class

bauom committed:
made CudaNewArray inherit from NumpyNewArray

framdani committed:
Fix deallocate printer (#1266)

Update the printer of deallocate to free the array object allocated by
one of the cuda's memory allocation APIs. Fixes pyccel/pyccel-cuda#8

 **Functions added:**
- `cuda_free_host` : frees the memory returned by cudaMallocHost().
- `cuda_free` : frees the array located on the device.
- `cuda_free_pointer` : frees a pointer to a cuda array object.

---------

Co-authored-by: EmilyBourne <louise.bourne@gmail.com>
Co-authored-by: bauom <40796259+bauom@users.noreply.github.com>

bauom committed:
fixing merge

bauom committed:
fixed basename issue && a test case

bauom committed:
changed attributes in CudaInternalVar, CudaSynchronize to class variables

bauom committed:
fixed docs string capitalisation

bauom committed:
removed cuda_constants empty dict

bauom committed:
removed unnecessary Asname call

bauom committed:
removed useless change

bauom committed:
updated epyccel docs string

bauom committed:
Update pyccel/decorators.py only importing array from numpy

Co-authored-by: EmilyBourne <louise.bourne@gmail.com>

bauom commented on pyccel/ast/cudaext.py:
```diff
@@ -0,0 +1,309 @@
+    __slots__ = ()
+    #--------------------------------------------------------------------------
+    @staticmethod
+    def _process_order(rank, order):
```
made `CudaNewArray` inherite from `NumpyNewArray`
[33e3439](https://github.com/pyccel/pyccel-cuda/pull/12/commits/33e34395fbc559e095b5dc3bbf084b0ba2587d04)

bauom committed:
added doc string to kernel decorator

bauom committed:
added kernel docs to tutorial/decorators.md

bauom committed:
improved doc string of kernel decorator

bauom commented on pyccel/decorators.py:
```diff
@@ -98,3 +100,7 @@ def allow_negative_index(f,*args):
         return f
     return identity
+
+def kernel(f):
```
[e9f9da9](https://github.com/pyccel/pyccel-cuda/pull/12/commits/e9f9da92087fac2a85214919a790bc7d25968e92)

EmilyBourne commented:
/bot run docs

EmilyBourne commented on pyccel/ast/cudaext.py:
```diff
@@ -48,22 +48,12 @@
-        if order not in ('C', 'F'):
-            raise ValueError('unrecognized order = {}'.format(order))
-        return order
+    pass
```
`pass` is not necessary. It is only needed when there is no code in the indented block.
E.g.
```python
class CudaNewArray(NumpyNewArray):

class A:
    ....
```
is problematic and needs `pass`, but:
```python
class CudaNewArray(NumpyNewArray):
     """ My docstring
     """

class A:
    ....
```
or:
```python
class CudaNewArray(NumpyNewArray):
    __slots__ = ()

class A:
    ....
```
is fine

EmilyBourne committed:
Merge branch 'pyccel:master' into master

EmilyBourne committed:
Merge remote-tracking branch 'origin/master' into cuda_master

EmilyBourne committed:
Merge branch 'pyccel:master' into master

EmilyBourne committed:
Merge branch 'pyccel:master' into master

EmilyBourne committed:
Merge branch 'pyccel:devel' into main

bauom committed:
removed unnecessary pass

bauom committed:
removed unused imports

bauom committed:
used f-string

bauom committed:
removed unused imports

bauom committed:
removed unused imports ccudacode

bauom committed:
fixed typo ??

bauom committed:
fixed indent

bauom committed:
using f-strings

EmilyBourne committed:
Merge remote-tracking branch 'origin/devel' into cuda_main

EmilyBourne committed:
Merge branch 'devel' into cuda_main_temp

EmilyBourne commented:
/bot run docs
